### DISCUSSION TITLE
how to use platform specific code?

### DISCUSSION THREAD

**Discussion by thawri1** [korlibs/korge#2018] [Q&A]
can use platform specific code in korge?
for example can  I use Yandex SDK in my games?
https://yandex.com/dev/games/doc/en/sdk/sdk-about

**Comment by soywiz** [comment]
You can use the `jsMain` source set and kotlin common with expect/actual.

Alternatively in common for dynamic code:

```
if (Platform.isJs) {
  // JSEval.
  // Dyn.global
}
``` 

But only in the JS source set you can define external classes. Which would be required for that.

By converting this code into kotlin you can load the yandex SDK:
```
var t = d.getElementsByTagName('script')[0];
var s = d.createElement('script');
s.src = 'https://yandex.ru/games/sdk/v2';
s.async = true;
t.parentNode.insertBefore(s, t);
s.onload = initSDK;
```  

Then you would have:

```
external val YaGames: dynamic
```

or you can provide externals for YaGames. I believe there is something to generate kotlin code from typescript definitions but don't remember the details:

```kotlin
external object YaGames {
  fun init(): Promise<dynamic>
}
``` 

Then you would (after the initSDK has been called) by converting this code into kotlin:

```
YaGames
    .init()
    .then(ysdk => {
        console.log('Yandex SDK initialized');
        window.ysdk = ysdk;
    });
```

**Comment by thawri1** [comment]
@soywiz 
thanks for your response.
I tried to do what you said but no success.
is there any template that has an Android and JS module and works with korge?

**Reply by soywiz** [reply]
Can you describe which kind of issues did you have to not have success?

**Reply by thawri1** [reply]
I'm sorry for taking your valuable time.
I have not worked with KMP before and tried to learn more about it.
Don't know how to use these codes.
I was hoping that there was a sample code so that I could learn how to use it

**Reply by thawri1** [reply]
what I try is:
adding KMP plugin:
`plugins {
    kotlin("multiplatform") version "1.9.20"
	alias(libs.plugins.korge)
}`

and then added this:
`kotlin {
    js {
        browser {
        }
        binaries.executable()
    }
}`

then I added JsMain to project like image below.
![Screenshot 2023-11-11 133910](https://github.com/korlibs/korge/assets/31119804/ffea5a77-9ce9-42a7-858b-2852d8bca1be)

but I don' know how to do after that.

**Reply by soywiz** [reply]
You define an expect in the `commonMain` folder:
* https://github.com/korlibs/show.korge.org/blob/5d1d1a239c5fcc3361485a42487b2b4356e84900/src/commonMain/kotlin/extension/Ext.kt#L13

Then you provide an implementation with `actual` in the targets you are using, for example `jsMain`:
* https://github.com/korlibs/show.korge.org/blob/5d1d1a239c5fcc3361485a42487b2b4356e84900/src/jsMain/kotlin/extension/JsExt.kt#L14

I'd recommend you to read more about it before using it in korge. Try to do a standalone project with a hello world expect/actual. Try external classes in JS, searching for Internet tutorials etc. there are better places than this for learning about this. You can use this as entry point: 
https://www.jetbrains.com/help/kotlin-multiplatform-dev/multiplatform-connect-to-apis.html

Then search for tutorials, ask in the Kotlin slack, etc.

After experimenting and understanding how it works, I bet this task will be easier.
